File: Sources/DLL/Arbre/Visiteurs/VisiteurSelectionnable.cpp
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurSelectionnable.cpp
/// @author Nam Lesage
/// @date 2016-09-29
/// @version 0.1
///
/// @addtogroup inf2990 INF2990
/// @{
///////////////////////////////////////////////////////////////////////////////

#include "VisiteurSelectionnable.h"

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::VisiteurSelectionnable(const std::string& type, const bool& selectionnable)
///
/// Constructeur de VisiteurSelectionnable.
///
///	@param[in] type : Type du noeud
///	@param[in] selectionnable : Si le noeud est sélectionnable ou non
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurSelectionnable::VisiteurSelectionnable(const std::string& type, const bool& selectionnable)
	: type_(type), selectionnable_(selectionnable)
{
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::~VisiteurSelectionnable()
///
/// Destructeur de VisiteurSelectionnable.
///
///	@param[in] Aucun
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurSelectionnable::~VisiteurSelectionnable() {
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::visiterAccelerateur(NoeudAccelerateur* noeud)
///
/// Cette fonction permet de visiter un accélérateur pour vérifier s'il est sélectionnable.
///
///	@param[in] noeud: Un accélérateur
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::visiterAccelerateur(NoeudAccelerateur* noeud) {
	implementationDefaut(noeud);
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::visiterMaillet(NoeudMaillet* noeud)
///
/// Cette fonction permet de visiter un maillet afin de l'implémenter par défaut.
///
///	@param[in] noeud: Un maillet
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::visiterMaillet(NoeudMaillet* noeud) {
	implementationDefaut(noeud);
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::visiterTable(NoeudTable* noeud)
///
/// Cette fonction permet de visiter une table afin de l'implémenter par défaut.
///
///	@param[in] noeud: Une table
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::visiterTable(NoeudTable* noeud) {
	implementationDefaut(noeud);
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::visiterPointControl(NoeudPointControl * noeud)
///
/// Cette fonction permet de visiter un point de contrôle afin de l'implémenter par défaut.
///
///	@param[in] noeud: Un point de contrôle
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::visiterPointControl(NoeudPointControl * noeud) {
	implementationDefaut(noeud);
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::visiterMur(NoeudMur* noeud)
///
/// Cette fonction permet de visiter un mur afin de l'implémenter par défaut.
///
///	@param[in] noeud: Un mur
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::visiterMur(NoeudMur* noeud) {
	implementationDefaut(noeud);
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::visiterPortail(NoeudPortail* noeud)
///
/// Cette fonction permet de visiter un portail afin de l'implémenter par défaut.
///
///	@param[in] noeud: Un portail
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::visiterPortail(NoeudPortail* noeud) {
	implementationDefaut(noeud);
}

void VisiteurSelectionnable::visiterRondelle(NoeudRondelle * noeud)
{
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::assignerType(const std::string & type)
///
/// Cette fonction permet d'assigner un type envoyé par paramètre.
///
///	@param[in] type: Type du noeud
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::assignerType(const std::string & type)
{
	type_ = type;
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::assignerSelectionnable(const bool & selectionnable)
///
/// Cette fonction permet de rendre un certain noeud sélectionnable.
///
///	@param[in] selectionnable: Si le noeud est sélectionnable ou non
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::assignerSelectionnable(const bool & selectionnable)
{
	selectionnable_ = selectionnable;
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurSelectionnable::implementationDefaut(NoeudAbstrait * noeud)
///
/// Cette fonction permet d'implémenter un noeud par défaut et de lui assigner
/// la caractéristique de sélectionnable.
///
///	@param[in] noeud: Un noeud
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurSelectionnable::implementationDefaut(NoeudAbstrait * noeud)
{
	if (noeud->obtenirType() == type_)
	{
		noeud->assignerEstSelectionnable(selectionnable_);
	}
}



///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////

File: AirHockeyLightClient/AirHockeyLightClient/Noyau/Arbre/Visiteurs/VisiteurSelectionnable.swift
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurSelectionnable.swift
/// @author Mikael Ferland et Pierre To
/// @date 2017-10-10
/// @version 0.1
///
/// @addtogroup log3900 LOG3900
/// @{
///////////////////////////////////////////////////////////////////////////////

///////////////////////////////////////////////////////////////////////////
/// @class VisiteurSelectionnable
/// @brief Permet de rendre selectionnable ou non un type de noeud
///
///
/// @author Mikael Ferland et Pierre
/// @date 2017-10-10
///////////////////////////////////////////////////////////////////////////
class VisiteurSelectionnable : VisiteurAbstrait {

    /// Type du noeud
    private var type: String

    /// Bool pour la caractéristique de sélection du noeud
    private var selectionnable: Bool = false
    
    /// Constructeur
    init(type: String, selectionnable: Bool) {
        self.type = type
        self.selectionnable = selectionnable
    }
    
    /// Assigne un type au noeud
    func assignerType(type: String) {
        self.type = type
    }
    
    /// Rend un noeud sélectionnable
    func assignerSelectionnable(selectionnable: Bool) {
        self.selectionnable = selectionnable
    }
    
    /// Visiter un accéléateur pour le rendre sélectionnable
    func visiterAccelerateur(noeud: NoeudAccelerateur) {
        self.implementationDefaut(noeud: noeud)
    }
    /// Visiter un maillet pour le rendre sélectionnable
    //virtual void visiterMaillet(NoeudMaillet* noeud);
    
    /// Visiter une table pour le rendre sélectionnable
    func visiterTable(noeud: NoeudTable) {
        self.implementationDefaut(noeud: noeud)
    }
    
    /// Visiter un point de contrôle pour le rendre sélectionnable
    func visiterPointControl(noeud: NoeudPointControl) {
        self.implementationDefaut(noeud: noeud)
    }
    
    /// Visiter un mur pour le rendre sélectionnable
    func visiterMur(noeud: NoeudMur) {
        self.implementationDefaut(noeud: noeud)
    }
    
    /// Visiter un portail pour le rendre sélectionnable
    func visiterPortail(noeud: NoeudPortail) {
        self.implementationDefaut(noeud: noeud)
    }

    //virtual void visiterRondelle(NoeudRondelle* noeud);
    
    /// Cette fonction permet d'implémenter un noeud par défaut et de lui assigner
    /// la caractéristique de sélectionnable.
    private func implementationDefaut(noeud: NoeudCommun) {
        if (noeud.obtenirType() == type)
        {
            noeud.assignerEstSelectionnable(selectionnable: selectionnable);
        }
    }
    
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////
